<a href="https://colab.research.google.com/github/alvarovillagranm/BTE320/blob/main/FINALIZED_In_Class_Assignments_BTE320_%E2%80%94_Alvaro_Villagran.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
"IN CLASS ASSIGNMENTS FOR ALVARO VILLAGRAN"

A company wants to calculate its Gross Profit as: Sales - Costs of Goods Sold

assume that cost of goods sold is 40%
The company wants to calclate taxes and net income:

prompt for a tax rate in %
To calculate taxes, multiply gross profit by tax rate, and to get net income you subtract taxes from gross profit

In [ ]:
Sales = input(("Enter Sales ($): "))
Sales = float(Sales)

cogs_pct = input("Enter Cost of Goods (%): ")
cogs_pct = float(cogs_pct)

cogs = Sales * (cogs_pct/100)

Gross_Profit = Sales - cogs

tax_rate = input("Enter Tax Rate (%): ")

tax_rate = float (tax_rate)

taxes = Gross_Profit * (tax_rate/100)

net_income = Gross_Profit - taxes
print("sales:", Sales)
print("Cost of goods:",cogs)
print("Gross Profit:", Gross_Profit)
print("Taxes:", taxes)
print("Net Income:", net_income)


Enter Sales ($): 10
Enter Cost of Goods (%): 2
Enter Tax Rate (%): 5
sales: 10.0
Cost of goods: 0.2
Gross Profit: 9.8
Taxes: 0.49000000000000005
Net Income: 9.31


INCOME STATEMENT CALCULATION

In [ ]:
Sales = input("Enter sales ($):")
Sales = float(Sales)

COGS = input("Enter COGS %: ")
COGS = float(COGS)


COGS = Sales * COGS / 100
Profit = Sales - COGS

Tax = input("Enter Tax %: ")
Tax = float(Tax)

Tax= ((100-Tax)/100) * Profit

print(f"Cost of Goods sold: ${COGS}\nGross Profit: ${Profit}\nNet Profit: ${Tax}")

NUMERICAL CAlCULATOR

In [ ]:
# Input
Equation_Type = input("Equation Symbol?")

First_Number= float(input("Enter first number: "))
Second_Number= float(input("Enter second number: "))

# Output
if Equation_Type == "+" :
    print(First_Number + Second_Number)
elif Equation_Type == "-" :
    print(First_Number - Second_Number)
elif Equation_Type == "**" :
  print(First_Number ** Second_Number)
elif Equation_Type == "/" :
  try:
    print(First_Number / Second_Number)
  except ZeroDivisionError:
    print("Cannot divide by zero")
elif Equation_Type == "//" :
  print(First_Number // Second_Number)
elif Equation_Type == "%" :
  print(First_Number % Second_Number)
elif Equation_Type == "*" :
  print(First_Number * Second_Number)
else :
  print("Invalid Operator")

Inventory Management

In [ ]:
import math

#Inputs
demand = int(input("Enter demand (units/year): "))
reorder = float(input("Enter reorder cost ($/order): "))
holding = float(input("Enter holding cost ($/unit/year): "))

# Process
eoq = (2*demand*reorder/holding) ** (1/2)

# Output
print(f"EOQ: {eoq} units")

Enter demand (units/year): 20
Enter reorder cost ($/order): 4
Enter holding cost ($/unit/year): 1
EOQ: 12.649110640673518 units


Inventory Management
Design a simple numerical calculator in Python. The user needs to provide three inputs: two numbers and the mathematical operator of their choice. The program should print the output on the screen.

In [ ]:
n1 = float(input("Enter number 1: "))
n2 = float(input("Enter number 2:"))
op = input("Enter operator (+,-,*,/,//,%,**)")

if op == '+':
  res = n1 + n2
elif op == "-":
  res = n1 - n2
elif op == "*":
  res = n1 * n2
elif op == "/":
  res = n1 / n2
elif op == "//":
  res = n1 // n2
elif op == "**":
  res = n1 ** n2
elif op == "%":
  res = n1 % n2
else:
  res = None
  print(f"Invalid operator")

print(f"result = {res}")

Enter number 1: 1
Enter number 2:2
Enter operator (+,-,*,/,//,%,**)/
result = 0.5


A parking garage charges  5plus 2.50 for eacg hour parked.

Write a program that prints the number of hours parked and the corresponding fee side by side, from 1 to 8 hours.

Example output: 1 7.5 2 10.0 3 12.5 4 15.0 5 17.50 6 20.0 7 22.5 8 25.0

Bonus: The minimum fee is  10andthemaximumfeeis 20

In [ ]:
for hour in range (1,9):
  fee = 5 + 2.5 * hour

  if fee < 10:
    fee = 10.0
  elif fee > 20:
    fee = 20.0
  print(hour, fee)


1 10.0
2 10.0
3 12.5
4 15.0
5 17.5
6 20.0
7 20.0
8 20.0


A movie theater manager is deciding how much to spend on advertisising for an upcoming movie. Each ticket is $10. Without any advertising, 20 people are forecasted to attend. Additional people may attend based on advertising spending, estimated to be:
  Additional attendees = 2 x advertisingDollars^1/2
In addition to advertising costs, there are fixed costs to run the theater of $200

Write a program to generate a table for different advertising amounts and profits. Example output(may differ slightly from your results):

In [ ]:
print("Ad ($) Profit ($)")
for ad in range(0,201,25):
  revenue = (20 + 2 * ad ** 0.5) * 10
  costs = 200 + ad
  profit = revenue - costs
  print(f"{ad}, \t {round(profit)}")

Ad ($) Profit ($)
0, 	 0
25, 	 75
50, 	 91
75, 	 98
100, 	 100
125, 	 99
150, 	 95
175, 	 90
200, 	 83


Append items to list iteratively (while loop)

starting with an empty list, desgin a program in python that populates the list by repeatedly asking the user for nurmerical inputs. The process stops when the user gives 0. Finally, the list should be displayed in descending order.

In [ ]:
numbers = []
num = float(input("Give a number: "))

while num != 0:
  numbers.append(num)
  num = float(input("Give another number: "))

numbers.sort(reverse=True)
print(numbers)

Give a number: 1
Give another number: 2
Give another number: 3
Give another number: 4
Give another number: 5
Give another number: 20
Give another number: 0
[20.0, 5.0, 4.0, 3.0, 2.0, 1.0]


Using python in colab: The classic hand game. Write a program that pays the fame against the computer for as long as the user wants. Start by allowing the user to choose whether they want to play using the appropriate input. As long as the use chooses 'yes', the game continues. Add realism by allowing the ocmputer to randomly choose its action. You can do thisusing the random library: import random, computer = random.choice(['rock', 'paper', 'scissors']

In [ ]:
import random

print("Welcome to Rock-Paper-Scissors!")

# Main game loop
while True:
    play = input("Do you want to play? (yes/no): ").strip().lower()
    if play != "yes":
        print("Thanks for playing! Goodbye.")
        break

    # Get user's choice
    user = input("Choose rock, paper, or scissors: ").strip().lower()
    if user not in ["rock", "paper", "scissors"]:
        print("Invalid choice. Please try again.")
        continue

    # Computer's random choice
    computer = random.choice(["rock", "paper", "scissors"])
    print(f"Computer chose: {computer}")

    # Decide winner
    if user == computer:
        print("It's a tie!")
    elif (user == "rock" and computer == "scissors") or \
         (user == "scissors" and computer == "paper") or \
         (user == "paper" and computer == "rock"):
        print("You win!")
    else:
        print("You lose!")

Welcome to Rock-Paper-Scissors!
Do you want to play? (yes/no): yes
Choose rock, paper, or scissors: rock
Computer chose: rock
It's a tie!
Do you want to play? (yes/no): no
Thanks for playing! Goodbye.


Product prices (dictionaries) A company sells two products, their prices are: P-01  9.95P−02 10.50 Write a program that sets up a dictionary with product IDs as keys and prices as their values. Prompt the user to enter a product ID, then display that product's price

sample run: Enter product ID: P-01 P-01 9.95 For your program with a dictionary of product IDs and their prices, refine it so that if the user enters a product ID not in the dictionary it display a message Sample run: Enter product ID: X-01 X-01 not found

In [ ]:
products = {
    'P-01': '$9.95',
    'P-02': '$10.50'
}

id = input('Enter product ID (P-0X): ')

if id in products:
  print(f'ID: {id}, Price: {products[id]}$')
else:
  print(f'{id} not found')

Enter product ID (p-0X): P-01
ID: P-01, Price: $9.95$


a parking garage charges 5 and $2.50 for each hour parked

write a prgram that prinkts the number of hours parked and the corresponding fee side-by-side Example output: 1 7.5 2 10.0 3 12.5 4 15.0 5 17.5 6 20.0 7 22.5 8 25.0

bonus: the minimum fee is 10 and the maximum fee is $20

Revisit the parking garage problem. wrap the iterative process into a function that takes as input parameter the number of parking hours and return the fee

bonus: use a second main(function) where the user dynamically assigns the number of hours from the keyboard. The function takes no input prameters but rather calls calcfee() amd returns fess



APPEND ITEMS

In [ ]:
L = []
num = float(input("Give me a number: "))

while num != 0:
  L.append(num)
  num = float(input("Enter a number: "))

L.sort(reverse= True)
print(L)

PRODUCT PRICES

In [ ]:
#Instructions
# a comapny sells 2 products: P-01 9.95, P-02 10.50

#input

products = {
    'p-01' : 9.95,
    'P-01'  : 9.95,
    'p-02' : 10.5,
    'P-02' : 10.5
}

ID = input('Enter Product ID ')

if ID in products:
  print(f'Product: {ID}, Price: {products[ID]}$')
else:
    print('Invalid Product Id')

In [ ]:
def parking(hours):
  fee = (hours * 2.5) + 5
  if fee < 10:
    print(f'Fee: $10')
  elif fee >= 10 and fee <= 20:
    print(f'Fee: ${fee}')
  elif fee > 20:
    print(f'Fee: $20')


def main():
  h = int(input('Enter Parking Hours: '))
  parking(h)

main()


Enter Parking Hours: 5
Fee: $17.5


Revisit the parking garage problem from before. wrap the iterative process into a function that takes as input parameter the number of parking hours and returns the fee.

Bonus: Use a second main() function where the user dynamically assigns the number of hours from the keyboard. This function takes no input parameters, but rather calls calcFee() and returns the fee.

In [ ]:
def calcFee(h):
 fee = 5 + 2.5 * h
 if fee < 10:
  return 10.0
 elif fee >= 10.0:
  return 20.0
 else:
  return fee

def main ():
  hours = int(input('Enter number of hours: '))
  fee = calcFee(hours)
  print(fee)

Recursive function: string reversal
- create a recursive function that takes as input a string and returns it reversed

- exaples:
  - revese('hello') -> 'olleh'
  - reverse('12345') -> '54321'

In [ ]:
def reverse(s):
  if len(s) > 1:
    #return s[-1] + reverse(s[:-1])
    #return reverse(s[1:]) + s[0]
    return s[-1] + reverse(s[1:-1]) + s[0]
  else:
    return s

reverse('hello')

'olleh'

A quality control manager would like to find the average weight of a list of weights. Prior experience indicates that the first two measure are atypical, and the practie is to exlude these from the average calculation.

Write a program that calulates the average of a list of numbers that given as inputs from the user, excluding the first two values.

For your program you must define and use the following functions:
- getValues(n): Prompt the use for n values, returns the list.
- calcAverage(alist): Calculates the average from the list.

Difference with Assignment II: use try-except to avoid invalid inputs(negative values, strings), invalid operations, etc.

In [ ]:
def getValues(n: int) -> list:
  '''
  Creates a list of weights. The eights are given by the user
  Input:
  - number of items in the list
  Output:
  - a list of weights
  '''
  items = []

  for i in range(n):
    try:
      item = float(input('Enter weight value: '))
    except Exception as e:
      print(e)
      item = float('nan')
    items.append(item)

  return items

def calcAverage(alist: list) -> float:
  '''
  Input:
  - A list of weights
  Output:
  - Average value; excludes the first two items
  '''
  try:
    return sum(alist[2:]) / len(alist[2:])
  except ZeroDivisionError:
    print('The list is empty!')
    return float('nan')

def main():
  num = int(input('Enter number of weights: '))

  if num <= 0:
    raise ValueError('num must be positive')

  weights = getValues(num)
  avg = calcAverage(weights)
  print(f'Average weight: {avg} lbs.')

main()

Enter number of weights: 5
Enter weight value: 2
Enter weight value: 2
Enter weight value: 5
Enter weight value: 0
Enter weight value: -1
Average weight: 1.3333333333333333 lbs.


LOAN & npv MINI TOOL KIT

In [ ]:
def loan_payment(principal: float, annual_rate: float, months: int) -> float:
    assert isinstance(principal, (int, float)), "principal must be numeric"
    assert isinstance(annual_rate, (int, float)), "annual_rate must be numeric"
    assert isinstance(months, int), "months must be an integer"

    if principal <= 0:
        raise ValueError("Principal must be > 0.")
    if annual_rate < 0:
        raise ValueError("Annual rate cannot be negative.")
    if months <= 0:
        raise ValueError("Number of months must be > 0.")

    r = annual_rate / 12.0
    n = months
    P = float(principal)

    if r == 0.0:
        return P / n
    else:
        return r * P / (1.0 - (1.0 + r) ** (-n))


if __name__ == "__main__":
    try:
        P = float(input("Enter principal: "))
        apr = float(input("Enter annual interest rate as a decimal: "))
        n = int(input("Enter number of months: "))
        payment = loan_payment(P, apr, n)
        print(f"Monthly payment: {payment:.2f}")
    except Exception as e:
        print(e)

Part A - compute simple ROI

In [ ]:
def roi(cost: float, revenue: float) -> float:
  assert cost > 0, "The cost must be positive!"
  assert isinstance(cost, (int, float)), "cost must be numerical"
  assert isinstance(revenue, (int, float)), "revenue must be numerical"

  return (revenue - cost) / cost

if __name__ == "__main__":
  c = float(input('Enter cost value: '))
  r = float(input('Enter revenue value: '))
  retInvest = roi(c, r)
  print(retInvest)


Enter cost value: 20
Enter revenue value: 30
0.5


Write a Python class and populate it with two attributes and methods:
- attributes: first/last names
- Method displ(): takes a Boolean input parameter; displays the attribute values either formally (last, first) or informally (first)
- Special method __str__()

In [ ]:
class Greeting:
  def __init__(self, first, last):
    self.first = first
    self.last = last

  def __str__(self):
    return "Instance of class Greeting"

  def display(self, reverse=True):
    if reverse:
      print(f'{self.last}, {self.first}')
    else:
      print(self.first)

name = Greeting('John', 'Doe')
print(name)
name.display()
name.display(reverse=False)

Instance of class Greeting
Doe, John
John


Create a class student that has
- three attributes: name (str), number (int), courses (list)
- two getter methods fro name and courses
- a setter method for name
- the special method __len__ that returns the length of list courses
- A method enroll() that appends a course to courses list if it not there. Basically a more elaborate setter method



In [ ]:
class Student:
  def __init__(self, name, number):
    self.__name = name
    self.__number = number
    self.__courses = []

  def getName(self):
    return self.__name

  def getCourses(self):
    return self.__courses

  def setName(self, newname):
   self.__name = newname

  def __len__(self):
    return len(self.__courses)

  def enroll(self, course):
    if course in self.__courses:
      print('Your have already enrolled in', course)
    else:
      self.__courses.append(course)

s = Student('Bob', 12345)
print(s.getName())
print(s.getCourses())
s.enroll('ABC123')
print(s.getCourses())
s.enroll('ABC123')
s.enroll('ABC567')
print(s.getCourses())
len(s)

Bob
[]
['ABC123']
Your have already enrolled in ABC123
['ABC123', 'ABC567']


2

Classes Point and Rectangle

Create a class Point that has:
- two coordinate attributes (x, y)
- A method distance() that calculates the distance between two instances of the class

Create a class rectangle that has:
 - six attributes: four Point instances and two attributes for width/height. The latter two are calvulated based on pairs of the four Point instances.
 - Three mthods for calculting width, height, area of the rectangle.

In [ ]:
class Point:
  def __init__(self, x, y):
    self.x = x
    self.y = y


  def distance(self, other):
    x_diff_sq = (self.x - other.x)**2
    y_diff_sq = (self.y - other.y)**2
    return (x_diff_sq + y_diff_sq)**0.5

class Rectangle:
  def __init__(self):
    self.p1 = Point(0,0)
    self.p2 = Point(1,0)
    self.p3 = Point(1,1)
    self.p4 = Point(0,1)
    self.width = self.p1.distance(self.p2)
    self.height = self.p1.distance(self.p4)

  def area(self):
   return self.width * self.height


sq = Rectangle()
sq.area()

1.0